contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,280 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult,
+    Uint128,
+};
+use cw2::set_contract_version;
+use cw20::Cw20ReceiveMsg;
+
+use crate::error::ContractError;
+use crate::msg::{
+    ExecuteMsg, InstantiateMsg, IsNameAvailableToRegisterResponse, LookUpDaoResponse,
+    LookUpNameResponse, QueryMsg, ReceiveMsg,
+};
+use crate::state::{Config, CONFIG, DAO_TO_NAME, NAME_TO_DAO, RESERVED_NAMES};
+
+const CONTRACT_NAME: &str = "crates.io:cw-name-registry";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let payment_token_address = deps.api.addr_validate(&msg.payment_token_address)?;
+    let admin = deps.api.addr_validate(&msg.admin)?;
+
+    if msg.payment_amount_to_register_name.is_zero() {
+        return Err(ContractError::InvalidPaymentAmount {});
+    }
+
+    let config = Config {
+        admin,
+        payment_token_address,
+        payment_amount_to_register_name: msg.payment_amount_to_register_name,
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(wrapped) => execute_receive(deps, env, info, wrapped),
+        ExecuteMsg::UpdateConfig {
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            new_payment_token_address,
+            new_admin,
+            new_payment_amount,
+        ),
+        ExecuteMsg::Revoke { name } => execute_revoke(deps, env, info, name),
+        ExecuteMsg::Reserve { name } => execute_reserve(deps, env, info, name),
+        ExecuteMsg::TransferReservation { name, dao } => {
+            execute_transfer_reservation(deps, env, info, name, dao)
+        }
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // We only take payments from our specified token
+    if info.sender != config.payment_token_address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // TODO: Is there a way we can verify this is a DAO
```

Comment:
Popping in to say that I'm unsure whether we _should_ verify that the registrant is a DAO. Memorializing something I said to @Callum-A yesterday: 

If we saw a bunch of degens snapping names, we would burn those names, shame those degens, and issue a patch as a DAO. In the meantime, why not support an entrepreneurial user snapping up a name for their DAO before they form said DAO? That sounds like something I would do.

Also, while requiring registrants be DAOs raises the barrier to name-snapping, a motivated spoofer will form a single-member DAO with the target name, set proposal length low, and register the name. The only real disincentive to gobbling names is the price we set, which the Name Registry DAO can adjust via governance proposal.

Since we have governance as a backstop, fewer permissions, and less complexity is IMO preferred.